extensions/spark/kyuubi-spark-connector-yarn/src/main/scala/org/apache/kyuubi/spark/connector/yarn/YarnAppPartitionReader.scala

```diff
@@ -0,0 +1,152 @@
+        case filters if filters.isEmpty => yarnClient.getApplications
+        // id => point query
+        // state => batch query
+        // type => in (a,b,c), batch query
+        case filters =>
+          filters.collectFirst {
+            case EqualTo("id", appId: String) => java.util.Collections.singletonList(
+                yarnClient.getApplicationReport(ApplicationId.fromString(appId)))
+            case EqualTo("state", state: String) =>
+              state match {
+                case validState if validYarnStateSet.contains(validState) =>
+                  yarnClient.getApplications(
+                    java.util.EnumSet.of(YarnApplicationState.valueOf(validState)))
+                case _ => Seq.empty[ApplicationReport].asJava
+              }
+            case EqualTo("type", appType: String) =>
+              yarnClient.getApplications(java.util.Collections.singleton(appType))
+            case In("state", states: Array[Any]) => yarnClient.getApplications(
+                java.util.EnumSet.copyOf(states
+                  .map(x => x.toString)
+                  .filter(x => validYarnStateSet.contains(x))
+                  .map(x =>
+                    YarnApplicationState.valueOf(x)).toList.asJava))
+            case In("type", types: Array[Any]) => yarnClient.getApplications(
+                types.map(x => x.toString).toSet.asJava)
+            case _ => yarnClient.getApplications()
+          }.get
+      }
+    val appSeq = applicationReports.asScala.filter(app =>
+      yarnAppPartition.filters
+        .forall(filter => maybeFilter(app, filter)))
+      .map(app => {
+        YarnApplication(
+          id = app.getApplicationId.toString,
+          appType = app.getApplicationType,
+          user = app.getUser,
+          name = app.getName,
+          state = app.getYarnApplicationState.name,
+          queue = app.getQueue,
+          attemptId = app.getCurrentApplicationAttemptId.toString,
+          submitTime = app.getSubmitTime,
+          launchTime = app.getLaunchTime,
+          startTime = app.getStartTime,
+          finishTime = app.getFinishTime,
+          trackingUrl = app.getTrackingUrl,
+          originalTrackingUrl = app.getOriginalTrackingUrl)
+      })
+    yarnClient.close()
+    appSeq
+  }
+
+  private def maybeFilter(app: ApplicationReport, filter: Filter): Boolean = {
+    filter match {
+      case EqualTo("id", appId: String) => app.getApplicationId.toString.equals(appId)
+      case EqualTo("state", appState: String) => app.getYarnApplicationState.name().equals(appState)
+      case EqualTo("type", appType: String) => app.getApplicationType.equals(appType)
+      case In("state", states) => states.map(x => x.toString)
+          .contains(app.getYarnApplicationState.name())
+      case In("type", types) => types.map(x => x.toString)
+          .contains(app.getApplicationType)
+      case _ => false
+    }
+  }
+}
+
+// Helper class to represent app
+case class YarnApplication(
```

Comment:
the class is not very helpful but would significantly increase memory consumption, let's convert the response data to InternelRow directly